inference/__init__.py

```diff
@@ -0,0 +1,492 @@
+        return W.flatten()
+    else:
+        return torch.cat(W).flatten()
+
+
+def updiv(x, y): return (x + y - 1) // y
+
+
+
+# Data structures
+class SPQRUncompressed:
+    m: int
+    n: int
+    bits: int
+    W: T
+    beta1: int
+    beta2: int
+    W_s: T
+    W_z: T
+    W_s_s: T
+    W_s_z: T
+    W_z_s: T
+    W_z_z: T
+    row_offsets: T
+    col_vals: T
+    dense_row_count: int
+    row_ids: T
+    buff0: T
+
+    @property
+    def nnz(self) -> int:
+        return self.col_vals.shape[0]
+
+    def __init__(self, m, n, bits, W, beta1, beta2, W_s, W_z, W_s_s, W_s_z, W_z_s, W_z_z, row_offsets, col_ids, values,
+                 in_perm=None, out_perm=None, sparse_compression_strategy: int = 0):
+        self.m = m
+        self.n = n
+        self.bits = bits
+        self.W = W
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.W_s = W_s
+        self.W_z = W_z
+        self.W_s_s = W_s_s
+        self.W_s_z = W_s_z
+        self.W_z_s = W_z_s
+        self.W_z_z = W_z_z
+        self.row_offsets = row_offsets
+        self.col_ids = col_ids
+        self.values = values
+        self.in_perm = in_perm
+        self.out_perm = out_perm
+        self.buff0 = allocate_compressed_buffers(m, n, beta1, beta2,  self.W.device)
+
+
+        if self.values.shape[0] != 0:
+            self.col_vals = values.view(torch.int16).to(torch.int64).bitwise_left_shift(16).bitwise_or(
+                col_ids.view(torch.int16).to(torch.int64)).to(torch.int32)
+        else:
+            self.col_vals = torch.zeros(0)
+
+        if sparse_compression_strategy == 0:
+            row_offsets_output = self.row_offsets
+            col_vals_output = self.col_vals
+        elif sparse_compression_strategy == 1:
+            row_offsets_output = self.row_offsets.diff().reshape((-1, 16)).max(axis=1).values * 16
+            row_offsets_output = row_offsets_output.cumsum(dim=0)
+            row_offsets_output = torch.cat((torch.tensor([0]), row_offsets_output)).to(dtype=torch.int32)
+            col_val_count = row_offsets_output[-1]
+            col_vals_output = torch.zeros(col_val_count, dtype=torch.int32)
+        else:
+            raise 'Invalid sparse compression strategy requested.'
+
+        spqr_cuda.tensor_compress_interleaved(m, n, bits, W, beta1, beta2, W_s, W_z, W_s_s, W_s_z, W_z_s, W_z_z,
+                                              self.row_offsets,
+                                              row_offsets_output,
+                                              self.col_vals,
+                                              col_vals_output,
+                                              self.buff0,
+                                              sparse_compression_strategy)
+        self.row_offsets = row_offsets_output
+        self.col_vals = col_vals_output
+
+
+class SPQRModule(torch.nn.Module):
+    def __init__(self, spqr_host: SPQRUncompressed):
+        super().__init__()
+        self.m = spqr_host.m
```

Comment:
consider renaming m and n into col row or something more readable (not that important)

Reply:
what is m and n in this class?

Reply:
Refactored this into QuantizedLinear to match HF naming convention with documentation.

Reply:
(for reference: m, n is the matrix height and width respectively).